### PR TITLE
Support converting HTML with nested list elements

### PR DESCRIPTION
**Pull request by hedgerwang**
For now, converting HTML with nested list elements is not properly supported.

For example, the following HTML 

```
<ol>
    <li>AAA</li>
    <li>AAA</li>
    <li>AAA
        <ol>
            <li>BBB</li>
            <li>BBB
                <ol>
                    <li>CCC</li>
                    <li>CCC</li>
                    <li>CCC</li>
                </ol>
            </li>
            <li>BBB</li>
        </ol>
    </li>
    <li>AAA</li>
    <li>AAA</li>
    <li>AAA
        <ol>
            <li>BBB</li>
            <li>BBB</li>
            <li>BBB

                <ol>
                    <li>CCC</li>
                    <li>CCC</li>
                    <li>CCC</li>
                </ol>
            </li>
        </ol>
    </li>
</ol>
```

will be rendered as 

---
image
---

Now that the inner list becomes flatten, which is not what we want.
But with this PR, it'll generate the right result

---
image
---

Note that this hasn't been an issue cause we've been only dealing HTML exported from Google Doc which never exports nest list elements.

For example, these nested lists:

---
image
---

Google Doc will exports HTML as 3 adjacent list elements.

```

<ol start="1">
    <li>AAA</li>
    <li>AAA</li>
</ol>
<ol style="margin-left: 24pt" start="1">
    <li>BBB</li>
    <li>BBB</li>
</ol>
<ol start="3">
    <li>AAA</li>
</ol>
```

Alternatively, MSWord exports  HTML as 2 nested list elements.

```
<ol>
    <li>AAA</li>
    <li>AAA
        <ol>
            <li>BBB</li>
            <li>BBB</li>
        </ol>
    </li>
    <li>AAA</li>
</ol>
```

Note that Google and MSWord exports the HTML that look visually same but structurally different, we'd have to deal with this special edge case.

### CODE REVIEW

**Comment by alextriana** [comment]
This seems legit, but it was admittedly a little weird to me to see us pull the list out and directly mutate the parent node as we iterate through the nested lists, as opposed to building up a new document fragment from scratch . My worry was that moving stuff "upward" is more confusing when the document itself is changing.

Can you perhaps briefly described why you mutated directly as opposed to building a frag?

**Comment by fresh-prince-of-prosemirror** [comment]
This seems legit, but it was admittedly a little weird to me to see us pull the list out and directly mutate the parent node as we iterate through the nested lists, as opposed to building up a new document fragment from scratch . My worry was that moving stuff "upward" is more confusing when the document itself is changing.

Can you perhaps briefly described why you mutated directly as opposed to building a frag?

**Comment by hedgerwang** [comment]
@fresh-prince-of-prosemirror : you're right. My original test HTML does work but the following HTML won't.

```
<ol>
    <li>AAA</li>
    <li>AAA</li>
    <li>AAA
        <ol>
            <li>BBB</li>
            <li>BBB
                <ol>
                    <li>CCC</li>
                    <li>CCC</li>
                    <li>CCC
                        <ol>
                            <li>DDD</li>
                            <li>DDD</li>
                            <li>DDD</li>
                        </ol>

                    </li>
                </ol>
            </li>
            <li>BBB</li>
        </ol>
    </li>
    <li>AAA</li>
    <li>AAA</li>
    <li>AAA
        <ol>
            <li>BBB</li>
            <li>BBB</li>
            <li>BBB

                <ol>
                    <li>CCC</li>
                    <li>CCC</li>
                    <li>CCC</li>
                </ol>
            </li>
        </ol>
    </li>
</ol>
```

I'll refactor my implementation and resolve the problem.
Thanks.

**Comment by hedgerwang** [comment]
re @freeman-lab : 

I changed my implementation and now I use a utility `HTMLMutator` which builds the new HTML for the new structure needed.

Please take another look when you have time.

Thanks.

File: src/HTMLMutator.js [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+  insertHTMLAfter(html: string, el: Element): HTMLMutator {
+    return this._insertHTML(html, 'after', el);
+  }
+
+  execute(): void {
+    const doc = this._doc;
+    const root = nullthrows(doc.body || doc.documentElement);
+    let newHtml = root.innerHTML;
+    this._htmls.forEach((html, token) => {
```
Comment (fresh-prince-of-prosemirror) [inline src/HTMLMutator.js]: `token` feels a little misleading here, because that second param is an index, right?

**Pull request merged**